.github/copilot-instructions.md

```diff
@@ -0,0 +1,211 @@
+└── lexer/            # Additional lexer modules
+    └── parser/       # Additional parser modules
+```
+
+### Configuration System
+quartofmt uses a hierarchical config lookup:
+1. Explicit `--config` path (errors if invalid)
+2. `.quartofmt.toml` or `quartofmt.toml` in current/parent directories  
+3. `~/.config/quartofmt/config.toml` (XDG)
+4. Built-in defaults (80 char width, auto line endings, reflow wrap)
+
+### Test Architecture  
+```
+tests/
+├── golden_cases.rs   # Main integration tests using test case directories
+├── cases/           # Input/expected output pairs (14 test scenarios)
+│   ├── wrap_paragraph/
+│   │   ├── input.qmd     # Raw input
+│   │   └── expected.qmd  # Expected formatted output
+│   └── ...
+└── format/          # Unit tests organized by feature
+    ├── code_cells.rs
+    ├── headings.rs
+    ├── math.rs
+    └── ...
+```
+
+### Workspace Structure
+```
+crates/
+└── quartofmt-wasm/   # WebAssembly bindings for web playground
+    ├── Cargo.toml
+    └── src/
+```
+
+### Build Configuration Files
+- `Cargo.toml`: Main project config, dependencies, binary definition
+- `rust-toolchain.toml`: Pins to stable Rust toolchain  
+- `Taskfile.yml`: Task runner commands (go-task) - NOT available in CI
+- `devenv.nix`: Nix development environment setup
+
+## CI/CD and Validation Pipeline
+
+### GitHub Actions Workflows
+1. **build-and-test.yml**: Main CI (Ubuntu/Windows/macOS, Rust stable)
+   - cargo build --verbose
+   - cargo test --verbose  
+   - cargo clippy -- -D warnings
+   - cargo fmt -- --check
+   - Security audit via rustsec/audit-check
+
+2. **release.yml**: Semantic release workflow
+   - Triggered manually
+   - Uses semantic-release with conventional commits
+
+3. **docs.yml**: Quarto documentation publishing to GitHub Pages
+
+### Pre-commit Validation
+The CI exactly replicates these commands - ensure all pass locally:
+```bash
+cargo build --verbose
+cargo test --verbose
+cargo clippy -- -D warnings
+cargo fmt -- --check
+```
+
+### Golden Test System
+The project uses snapshot testing via `tests/golden_cases.rs`:
+- Each `tests/cases/*` directory contains `input.qmd` and `expected.qmd`
+- Tests verify formatting is idempotent (format twice = format once)
+- Use `UPDATE_EXPECTED=1 cargo test` to update expected outputs (BE CAREFUL)
+
+## Key Development Facts
+
+### Dependencies
+- **clap**: CLI argument parsing with derive macros
+- **rowan**: Red-green tree for CST (Concrete Syntax Tree)  
+- **regex**: Pattern matching for lexing
+- **textwrap**: Text wrapping functionality
+- **toml**: Configuration file parsing
+- **serde**: Serialization for config structs
+
+### Testing Approach  
+- Unit tests embedded in source modules (34 tests)
+- Integration tests in `tests/format/` (39 tests) 
+- Golden tests comparing input/expected pairs (1 comprehensive test covering 14 scenarios)
+- Property: formatting is idempotent
+
+### Formatting Rules
```

Comment:
Formatting will to a large degree be configurable through the configuration file.